### PR TITLE
Java Flavour of PostgreSQL guide

### PR DESCRIPTION
**Pull request by agoerler**

### CODE REVIEW

File: guides/databases-postgres.md [hunk 1]
```diff
@@ -1,29 +1,64 @@
...
 
+<div markdown="1" class="impl node">
+
 Run this to use [PostgreSQL](https://www.postgresql.org/) for production:
 
+</div>
+
+<div markdown="1" class="impl java">
+
+To run CAP Java on PostgreSQL, add a Maven dependency to the PostgreSQL JDBC driver:
```
Comment (mofterdinger) [inline guides/databases-postgres.md:37]:
```suggestion
To run CAP Java on PostgreSQL, add a Maven dependency to the PostgreSQL JDBC driver into `srv/pom.xml`:
```

File: guides/databases-postgres.md [hunk 2]
```diff
@@ -73,17 +112,74 @@ You can use Docker to run a PostgreSQL database locally as follows:
...
+
+### Service Bindings for CDS tooling {.impl .java}
 
+<div markdown="1" class="impl java">
+
+If you use `cds deploy` deploy to directly [deploy to a PostgreSQL database](#using-cds-deploy) you need to additionally configure the service bindings (connection data) for the CDS tooling. Add them to your private `.cdsrc-private.json`. If you run postgres in a [docker container](#using-docker) use:
+
+::: code-group
+
+```json [~/.cdsrc.json]
```
Comment (mofterdinger) [inline guides/databases-postgres.md:162]: Should it be `~/.cdsrc.json` or `~/.cdsrc-private.json` ? Both names are mentioned, not sure which one to use.

File: guides/databases-postgres.md [hunk 3]
```diff
@@ -73,17 +112,74 @@ You can use Docker to run a PostgreSQL database locally as follows:
...
+        "user": "postgres",
+        "password": "postgres",
+        "database": "postgres"
+      }
+    }
+  }
+}
+```
+
+:::
```
Comment (agoerler) [inline guides/databases-postgres.md:178]:
```suggestion
```json
{
  "requires": {
    "db": {
      "kind": "postgres",
      "credentials": {
        "host": "localhost", "port": 5432,
        "user": "postgres",
        "password": "postgres",
        "database": "postgres"
      }
    }
  }
}
```
```

File: guides/databases-postgres.md (hunk 2 above)
Comment (agoerler) [reply guides/databases-postgres.md:162]: Just removed the file name from the code group

File: guides/databases-postgres.md [hunk 4]
```diff
@@ -71,17 +112,70 @@ You can use Docker to run a PostgreSQL database locally as follows:
...
+  driver-class-name: org.postgresql.Driver
+```
+
+Learn more about the [configuration of a PostgreSQL database](../java/persistence-services#configure-postgresql){ .learn-more}
+
+### Service Bindings for CDS tooling {.impl .java}
+
+<div markdown="1" class="impl java">
+
+If you use `cds deploy` deploy to directly [deploy to a PostgreSQL database](#using-cds-deploy) you need to additionally configure the service bindings (connection data) for the CDS tooling. Add them to your private `.cdsrc-private.json`. If you run postgres in a [docker container](#using-docker) use:
```
Comment (beckermarc) [inline guides/databases-postgres.md:158]: As shown in the keynote today, maybe we can directly leverage `cds deploy --profile pg` which brings these credentials out-of-the-box?

File: guides/databases-postgres.md [hunk 5]
```diff
@@ -71,17 +112,70 @@ You can use Docker to run a PostgreSQL database locally as follows:
...
+If you run the PostgreSQL database in a [docker container](#using-docker) your connection data might look like this:
+
+```yaml
+spring.datasource:
+  url: jdbc:postgresql://localhost:5432/postgres
+  username: postgres
+  password: postgres
+  driver-class-name: org.postgresql.Driver
+```
+
```
Comment (mofterdinger) [inline guides/databases-postgres.md:151]:
```suggestion
To start the application with the new profile `postgres`, the `spring-boot-maven-plugin` can be used: `mvn spring-boot:run -Dspring-boot.run.profiles=postgres`.
```

Comment (mofterdinger) [reply guides/databases-postgres.md:151]: Not sure if we should describe this here or not. Just a suggestion.

File: guides/databases-postgres.md [hunk 6]
```diff
@@ -336,11 +428,102 @@ We can use `cds deploy` with option `--dry` to simulate and inspect how things w
 
    > **Note:** In case of SQLite, ALTER TYPE commands are neither necessary nor supported, as SQLite is essentially typeless.
 
+## Deployment Using Liquibase  { .impl .java }
+
+You can also use [Liquibase](https://www.liquibase.org/) to control when, where, and how database changes are deployed. As Liquibase allows to define database changes [in an SQL file](https://docs.liquibase.com/change-types/sql-file.html), you can easily leverage `cds deploy` to generate DDL scripts to be used by Liquibase.
+
+Add a Maven dependency to Liquibase:
```
Comment (mofterdinger) [inline guides/databases-postgres.md:435]:
```suggestion
Add a Maven dependency to Liquibase in `srv/pom.xml`:
```

File: guides/databases-postgres.md [hunk 7]
```diff
@@ -338,11 +432,102 @@ We can use `cds deploy` with option `--dry` to simulate and inspect how things w
 
    > **Note:** In case of SQLite, ALTER TYPE commands are neither necessary nor supported, as SQLite is essentially typeless.
 
+## Deployment Using Liquibase  { .impl .java }
```
[the remark this answers is not in the document]
Comment (renejeglinsky) [reply guides/databases-postgres.md:431]: Spoke yesterday with Daniel, he has no objections.

**Comment by beckermarc** [comment]
I added my changes to reflect the new `cds-feature-postgresql` and the service binding auto-configuration it includes

File: guides/databases-postgres.md [hunk 8]
```diff
@@ -336,11 +422,102 @@ We can use `cds deploy` with option `--dry` to simulate and inspect how things w
...
+
+```yml
+databaseChangeLog:
+   - changeSet:
+       id: 1
+       author: me
+       changes:
+       - sqlFile:
+           dbms: postgresq
+           path: changelog/v1/schema.sql
```
Comment (mofterdinger) [inline guides/databases-postgres.md:450]:
```suggestion
           path: db/changelog/v1/schema.sql
```

File: guides/databases-postgres.md [hunk 9]
```diff
@@ -336,11 +422,102 @@ We can use `cds deploy` with option `--dry` to simulate and inspect how things w
...
+```
+
+Use `cds deploy` to create the _v1/schema.sql_ file:
+
+```sh
+cds deploy --dry > srv/src/main/resources/changelog/v1/schema.sql
```
Comment (mofterdinger) [inline guides/databases-postgres.md:456]:
```suggestion
cds deploy --dry > srv/src/main/resources/db/changelog/v1/schema.sql
```

File: guides/databases-postgres.md [hunk 10]
```diff
@@ -336,11 +422,102 @@ We can use `cds deploy` with option `--dry` to simulate and inspect how things w
...
+```
+
+In the _application.yml_ file, configure Spring to use Liquibase for schema evolution:
+
+```yml
+spring:
```
Comment (mofterdinger) [inline guides/databases-postgres.md:462]: If you use the default location `db/changelog/db.changelog-master.yml` this configuration is not required.

File: guides/databases-postgres.md [hunk 11]
```diff
@@ -336,11 +422,102 @@ We can use `cds deploy` with option `--dry` to simulate and inspect how things w
...
    > **Note:** In case of SQLite, ALTER TYPE commands are neither necessary nor supported, as SQLite is essentially typeless.
 
+## Deployment Using Liquibase  { .impl .java }
+
+You can also use [Liquibase](https://www.liquibase.org/) to control when, where, and how database changes are deployed. As Liquibase allows to define database changes [in an SQL file](https://docs.liquibase.com/change-types/sql-file.html), you can easily leverage `cds deploy` to generate DDL scripts to be used by Liquibase.
+
+Add a Maven dependency to Liquibase in `srv/pom.xml`:
+
+```xml
+<dependency>
```
Comment (mofterdinger) [inline guides/databases-postgres.md:432]: Dependency could be runtime scope, if liquibase API isn't directly used in custom code.

File: java/persistence-services.md [hunk 12]
```diff
@@ -106,7 +106,7 @@ Service bindings of type *service-manager* and, in a Spring-based application, *
 
 ### PostgreSQL { #configure-postgresql }
 
-PostgreSQL can be configured when running locally as well as when running productively in the cloud. However, for a cloud deployment no auto-configuration is available. 
+PostgreSQL can be configured when running locally as well as when running productively in the cloud. Similar to HANA, the datasource is auto-configured based on available service bindings, if the dependency `cds-feature-postgresql` is available on the classpath.
```
Comment (agoerler) [inline java/persistence-services.md:109]:
```suggestion
PostgreSQL can be configured when running locally as well as when running productively in the cloud. Similar to HANA, the datasource is auto-configured based on available service bindings, if the feature `cds-feature-postgresql` is installed.
```

installed | configured | added ?

File: guides/databases-postgres.md [hunk 13]
```diff
@@ -336,11 +422,102 @@ We can use `cds deploy` with option `--dry` to simulate and inspect how things w
...
+Once you are ready to release an initial version of your database schema, you can create a DDL file that defines the initial database schema. Firstly create subfolder `changelog` under `srv/src/main/resources`. Here you place the Liquibase [change log](https://docs.liquibase.com/concepts/changelogs/home.html) and the DDL scripts for the schema versions. The change log is described by the _changelog/changelog.yml_ file:
+
+```yml
+databaseChangeLog:
+   - changeSet:
+       id: 1
+       author: me
+       changes:
+       - sqlFile:
+           dbms: postgresq
```
Comment (mofterdinger) [inline guides/databases-postgres.md:449]:
```suggestion
           dbms: postgresql
```

File: guides/databases-postgres.md [hunk 14]
```diff
@@ -336,11 +422,102 @@ We can use `cds deploy` with option `--dry` to simulate and inspect how things w
...
+           path: changelog/v1/schema.sql
```
Comment (mofterdinger) [inline guides/databases-postgres.md:502]:
```suggestion
           dbms: postgresql
           path: db/changelog/v1/schema.sql
```

File: guides/databases-postgres.md [hunk 15]
```diff
@@ -336,11 +422,102 @@ We can use `cds deploy` with option `--dry` to simulate and inspect how things w
...
+   - changeSet:
+       id: 2
+       author: me
+       changes:
+       - sqlFile:
+           dbms: postgresq
+           path: changelog/v2/delta.sql
```
Comment (mofterdinger) [inline guides/databases-postgres.md:509]:
```suggestion
           dbms: postgresql
           path: db/changelog/v2/delta.sql
```

File: guides/databases-postgres.md [hunk 16]
```diff
@@ -336,11 +422,102 @@ We can use `cds deploy` with option `--dry` to simulate and inspect how things w
...
+```
+
+Don't forget to store the CSN file, which corresponds to this schema version:
 
+```sh
+cds deploy --model-only --dry > srv/src/main/resources/changelog/v2/csn.json
```
Comment (mofterdinger) [inline guides/databases-postgres.md:515]:
```suggestion
cds deploy --model-only --dry > srv/src/main/resources/db/changelog/v2/csn.json
```

File: guides/databases-postgres.md [hunk 17]
```diff
@@ -336,11 +422,102 @@ We can use `cds deploy` with option `--dry` to simulate and inspect how things w
...
+
+### Next Schema Versions
+
+If changes of the CDS model require changes on the database, you can create a new change set that captures the necessary changes.
 
+Use `cds deploy` to compute the delta DDL script based on the previous model versions (_v1/csn.json_) and the current model. Write the diff into a _v2/delta.sql_ file:
 
+```sh
+cds deploy --dry --delta-from srv/src/main/resources/changelog/v1/csn.json >
+                              srv/src/main/resources/changelog/v2/delta.sql
```
Comment (mofterdinger) [inline guides/databases-postgres.md:489]:
```suggestion
cds deploy --dry --delta-from srv/src/main/resources/db/changelog/v1/csn.json >
                              srv/src/main/resources/db/changelog/v2/delta.sql
```

File: guides/databases-postgres.md [hunk 18]
```diff
@@ -336,11 +422,102 @@ We can use `cds deploy` with option `--dry` to simulate and inspect how things w
...
+```yml
+spring:
+    liquibase:
+        change-log: classpath:/changelog/changelog.yml
+```
+
+Finally, store the CSN file, which corresponds to this schema version:
+
+```sh
+cds deploy --model-only --dry > srv/src/main/resources/changelog/v1/csn.json
```
Comment (mofterdinger) [inline guides/databases-postgres.md:470]:
```suggestion
cds deploy --model-only --dry > srv/src/main/resources/db/changelog/v1/csn.json
```

File: guides/databases-postgres.md [hunk 19]
```diff
@@ -498,21 +498,21 @@ databaseChangeLog:
        author: me
        changes:
        - sqlFile:
-           dbms: postgresq
-           path: changelog/v1/schema.sql
+           dbms: postgresql
```
[the remark this answers is not in the document]
Comment (mofterdinger) [reply guides/databases-postgres.md:504]: yes, we can try the relative paths, not sure if they work here.

File: guides/databases-postgres.md [hunk 20]
```diff
@@ -337,11 +423,106 @@ We can use `cds deploy` with option `--dry` to simulate and inspect how things w
...
+cds deploy --profile pg --dry > srv/src/main/resources/db/changelog/v1/schema.sql
+```
+
+In the _application.yml_ file, configure Spring to use Liquibase for schema evolution:
+
+```yml
+spring:
+    liquibase:
+        change-log: classpath:/changelog/changelog.yml
+```
```
Comment (mofterdinger) [inline guides/databases-postgres.md:468]: Remove unnecessary configuration if we use default location of liquibase master changelog.

```suggestion
```

**Comment by mofterdinger** [comment]
Is it possible to get a preview of this PR to see the final result ?

File: guides/databases-postgres.md [hunk 21]
```diff
@@ -1,27 +1,64 @@
...
+</div>
+
 ```sh
 npm add @cap-js/postgres
 ```
 
-### Auto-Wired Configuration
+<div markdown="1" class="impl java">
+
+This allows to use the `cds deploy` command to [deploy](#using-cds-deploy) to a PostgreSQL database or to create a [create DDL script](#deployment-using-liquibase) for PostgreSQL.
```
Comment (beckermarc) [inline guides/databases-postgres.md:57]:
```suggestion
This allows to use the `cds deploy` command to [deploy](#using-cds-deploy) to a PostgreSQL database or to [create a DDL script](#deployment-using-liquibase) for PostgreSQL.
```

File: guides/databases-postgres.md [hunk 22]
```diff
@@ -43,7 +80,11 @@ Output:
 
 [See also the general information on installing database packages](databases#setup-configuration){.learn-more}
 
+## Provisoning a Database Instance
+
+Productively, you can leverage the [PostgreSQL on SAP BTP, hyperscaler option](https://discovery-center.cloud.sap/serviceCatalog/postgresql-hyperscaler-option) to connect the a PostgreSQL offering from the cloud provider.
```
Comment (beckermarc) [inline guides/databases-postgres.md:85]:
```suggestion
Productively, you can leverage the [PostgreSQL on SAP BTP, hyperscaler option](https://discovery-center.cloud.sap/serviceCatalog/postgresql-hyperscaler-option) to connect to a PostgreSQL offering from the cloud provider.
```

**Comment by vmikhailenko** [comment]
> Is it possible to get a preview of this PR to see the final result ?

I use `npm run start` locally.

File: guides/databases-postgres.md [hunk 23]
```diff
@@ -71,17 +112,64 @@ You can use Docker to run a PostgreSQL database locally as follows:
...
+  datasource:
+    url: jdbc:postgresql://localhost:5432/postgres
+    username: postgres
+    password: postgres
+    driver-class-name: org.postgresql.Driver
+```
+To start the application with the new profile `postgres-docker`, the `spring-boot-maven-plugin` can be used: `mvn spring-boot:run -Dspring-boot.run.profiles=postgres-docker`.
+Learn more about the [configuration of a PostgreSQL database](../java/persistence-services#postgresql-1){ .learn-more}
+
+You can leverage Spring Boot 3.1 improved testcontainers support](https://spring.io/blog/2023/06/23/improved-testcontainers-support-in-spring-boot-3-1) to create PostgreSQL containers on the fly for local development or testing purposes.
```
Comment (beckermarc) [inline guides/databases-postgres.md:146]:
```suggestion
You can also leverage [Testcontainers support](https://spring.io/blog/2023/06/23/improved-testcontainers-support-in-spring-boot-3-1) introduced in Spring Boot 3.1 to create PostgreSQL containers on the fly for local development or testing purposes.
```

File: guides/databases-postgres.md [hunk 24]
```diff
@@ -71,17 +112,64 @@ You can use Docker to run a PostgreSQL database locally as follows:
...
+
+### Service Bindings for CDS tooling {.impl .java}
+
+<div markdown="1" class="impl java">
+
+If you use `cds deploy` deploy to directly [deploy to a PostgreSQL database](#using-cds-deploy) you need to additionally configure the service bindings (connection data) for the CDS tooling. Add them to your private `.cdsrc-private.json`. If you run postgres in a [docker container](#using-docker) use:
```
Comment (beckermarc) [inline guides/databases-postgres.md:152]:
```suggestion
If you use `cds deploy` to directly [deploy to a PostgreSQL database](#using-cds-deploy) you need to additionally configure the service bindings (connection data) for the CDS tooling. Add them to your private `.cdsrc-private.json`. If you run postgres in a [docker container](#using-docker) use:
```

File: guides/databases-postgres.md [hunk 25]
```diff
@@ -337,11 +424,98 @@ We can use `cds deploy` with option `--dry` to simulate and inspect how things w
...
+
+```xml
+<dependency>
+    <groupId>org.liquibase</groupId> 
+    <artifactId>liquibase-core</artifactId>
+    <scope>runtime</scope>
+</dependency>
+```
+
+Once `liquibase-core` is on the classpath, [Spring runs database migrations](https://docs.spring.io/spring-boot/docs/current/reference/html/howto.html#howto.data-initialization.migration-tool.liquibase) for application startup and before your tests run.
```
Comment (beckermarc) [inline guides/databases-postgres.md:441]:
```suggestion
Once `liquibase-core` is on the classpath, [Spring runs database migrations](https://docs.spring.io/spring-boot/docs/current/reference/html/howto.html#howto.data-initialization.migration-tool.liquibase) automatically on application startup and before your tests run.
```

File: guides/databases-postgres.md [hunk 26]
```diff
@@ -337,11 +424,98 @@ We can use `cds deploy` with option `--dry` to simulate and inspect how things w
...
+       changes:
+       - sqlFile:
+           dbms: postgresql
+           path: db/changelog/v1/model.sql
+```
+
+Use `cds deploy` to create the _v1/model.sql_ file:
+
+```sh
+cds deploy --profile pg --dry > srv/src/main/resources/db/changelog/v1/model.sql
```
Comment (beckermarc) [inline guides/databases-postgres.md:461]:
```suggestion
cds deploy --to postgres --dry > srv/src/main/resources/db/changelog/v1/model.sql
```

File: guides/databases-postgres.md [hunk 27]
```diff
@@ -337,11 +424,98 @@ We can use `cds deploy` with option `--dry` to simulate and inspect how things w
...
 
+### ② Schema Evolution { #schema-evolution-with-liquibase }
 
+If changes of the CDS model require changes on the database, you can create a new change set that captures the necessary changes.
 
+Use `cds deploy` to compute the delta DDL script based on the previous model versions (_v1/model.csn_) and the current model. Write the diff into a _v2/delta.sql_ file:
 
-## Migration
+```sh
+cds deploy --dry --delta-from srv/src/main/resources/db/changelog/v1/model.csn >
```
Comment (beckermarc) [inline guides/databases-postgres.md:484]:
```suggestion
cds deploy --to postgres --dry --delta-from srv/src/main/resources/db/changelog/v1/model.csn >
```

File: guides/databases-postgres.md (hunk 26 above)
Comment (mofterdinger) [reply guides/databases-postgres.md:461]: --to progress is the old and outdated profile as far as I know. I got this information from Simon from the compiler team.

**Review by beckermarc — APPROVED** [review]

**Comment by beckermarc** [comment]
@smahati Can you please review this PR, as it is relevant for the July release?

**Review by smahati — APPROVED** [review]

**Comment by smahati** [comment]
@beckermarc I'll merge this tomorrow but if you can and have a chance please take a look at my changes. Thanks.

File: guides/databases-postgres.md [hunk 28]
```diff
@@ -371,7 +371,7 @@ We can use `cds deploy` with option `--dry` to simulate and inspect how things w
    cds deploy --dry --model-only > cds-model.csn
    ```
 
-2. Make changes to your models, for example to *[cap/samples/bookshop/db/schema.cds](https://github.com/SAP-samples/cloud-cap-samples/blob/main/bookshop/db/schema.cds)*:
+2. Change your models, for example to *[cap/samples/bookshop/db/schema.cds](https://github.com/SAP-samples/cloud-cap-samples/blob/main/bookshop/db/schema.cds)*:
```
Comment (beckermarc) [inline guides/databases-postgres.md:374]:
```suggestion
2. Change your models, for example in *[cap/samples/bookshop/db/schema.cds](https://github.com/SAP-samples/cloud-cap-samples/blob/main/bookshop/db/schema.cds)*:
```

File: guides/databases-postgres.md [hunk 29]
```diff
@@ -429,11 +429,11 @@ We can use `cds deploy` with option `--dry` to simulate and inspect how things w
 
    :::
 
-   > **Note:** In case of SQLite, ALTER TYPE commands are neither necessary nor supported, as SQLite is essentially typeless.
+   > **Note:** If you use SQLite, ALTER TYPE commands are not necessary and so, are not supported, as SQLite is essentially typeless.
 
 ## Deployment Using Liquibase  { .impl .java }
 
-You can also use [Liquibase](https://www.liquibase.org/) to control when, where, and how database changes are deployed. As Liquibase allows to define database changes [in an SQL file](https://docs.liquibase.com/change-types/sql-file.html), you can easily leverage `cds deploy` to generate DDL scripts to be used by Liquibase.
+You can also use [Liquibase](https://www.liquibase.org/) to control when, where, and how database changes are deployed. Liquibase lets you define database changes [in an SQL file](https://docs.liquibase.com/change-types/sql-file.html), use `cds deploy` to quickly generate DDL scripts which can used by Liquibase.
```
Comment (beckermarc) [inline guides/databases-postgres.md:436]:
```suggestion
You can also use [Liquibase](https://www.liquibase.org/) to control when, where, and how database changes are deployed. Liquibase lets you define database changes [in an SQL file](https://docs.liquibase.com/change-types/sql-file.html), use `cds deploy` to quickly generate DDL scripts which can be used by Liquibase.
```

**Review by beckermarc — APPROVED** [review]

**Pull request merged**